//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by AsyncGenerator.
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------


using System;
using System.Collections;

using NUnit.Framework;

namespace NHibernate.Test.NHSpecificTest.NH940
{
	using System.Threading.Tasks;
	[TestFixture]
	public class NH940FixtureAsync : BugTestCase
	{
		[Test]
		public async Task BugAsync()
		{
			A a = new A();
			B b = new B();
			a.B = b;

			using (ISession s = OpenSession())
			using (ITransaction t = s.BeginTransaction())
			{
				await (s.SaveAsync(a));
				await (t.CommitAsync());
				s.Close();
			}

			using (ISession s = OpenSession())
			{
				IList l = await (s.CreateCriteria(typeof(A)).ListAsync());
				try
				{
					((A)l[0]).Execute();
					Assert.Fail("Should have thrown MyException");
				}
				catch (MyException)
				{
					// OK
				}
				catch (Exception e)
				{
					Assert.Fail("Should have thrown MyException, thrown {0} instead", e);
				}
			}

			using (ISession s = OpenSession())
			using (ITransaction t = s.BeginTransaction())
			{
				await (s.DeleteAsync(a));
				await (t.CommitAsync());
				s.Close();
			}
		}
	}
}